### PR TITLE
Enforce quota on segment upload

### PR DESCRIPTION
**Pull request by antumbde**
Enforce storage quota configured in table config on segment upload
request. This is no-op if the storage quota configuration is not
set in the table config.

Testing: UT + manual

### CODE REVIEW

File: pinot-controller/src/main/java/com/linkedin/pinot/controller/api/restlet/resources/PinotSegmentUploadRestletResource.java [hunk 1]
```diff
@@ -414,13 +415,23 @@ private String getDownloadUri(FileUploadType uploadType, Series headers, Represe
...
   private Representation uploadSegment(File indexDir, File dataFile, String downloadUrl)
       throws ConfigurationException, IOException, JSONException {
     final SegmentMetadata metadata = new SegmentMetadataImpl(indexDir);
     final File tableDir = new File(baseDataDir, metadata.getTableName());
     File segmentFile = new File(tableDir, dataFile.getName());
+    StorageQuotaChecker.QuotaCheckerResponse quotaResponse = checkStorageQuota(indexDir, metadata);
+    if (!quotaResponse.isSegmentWithinQuota) {
+      // this is not an "error" hence we don't increment segment upload errors
+      setStatus(Status.CLIENT_ERROR_FORBIDDEN);
+      StringRepresentation repr = new StringRepresentation("{\"error\" : \"" + quotaResponse.reason + "\"}");
```
Comment (mcvsubbu) [inline pinot-controller/src/main/java/com/linkedin/pinot/controller/api/restlet/resources/PinotSegmentUploadRestletResource.java]: Please log an INFO here, with the quota, current size, expected increased size due to this segment, and of course table name.

Comment (mcvsubbu) [inline pinot-controller/src/main/java/com/linkedin/pinot/controller/api/restlet/resources/PinotSegmentUploadRestletResource.java]: Looks like you have this info in quota checker response, so maybe just log that as in INFO?

Basically, log all info that we need to explain why quota exceeded.thanks

File: pinot-controller/src/test/java/com/linkedin/pinot/controller/validation/StorageQuotaCheckerTest.java [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+
+  @BeforeClass
+  public void setUp() {
+    tableSizeReader = mock(TableSizeReader.class);
+    tableConfig = mock(AbstractTableConfig.class);
+    quotaConfig = mock(QuotaConfig.class);
+    validationConfig = mock(SegmentsValidationAndRetentionConfig.class);
+    when(tableConfig.getValidationConfig()).thenReturn(validationConfig);
+    when(validationConfig.getReplicationNumber()).thenReturn(2);
+    TEST_DIR.mkdirs();
```
Comment (mcvsubbu) [inline pinot-controller/src/test/java/com/linkedin/pinot/controller/validation/StorageQuotaCheckerTest.java]: deleteOnExit()?

**Comment by mcvsubbu** [comment]
Otherwise, LGTM

File: pinot-controller/src/main/java/com/linkedin/pinot/controller/api/restlet/resources/PinotSegmentUploadRestletResource.java (hunk 1 above)
Comment (antumbde) [inline pinot-controller/src/main/java/com/linkedin/pinot/controller/api/restlet/resources/PinotSegmentUploadRestletResource.java]: Added info log with table and segment name and reason that looks like "Estimated size: 79807673 bytes exceeds the configured quota of 62914560 (bytes). Incoming segment size: 79807674 (bytes)"

File: pinot-controller/src/test/java/com/linkedin/pinot/controller/validation/StorageQuotaCheckerTest.java (hunk 2 above)
Comment (antumbde) [inline pinot-controller/src/test/java/com/linkedin/pinot/controller/validation/StorageQuotaCheckerTest.java]: There is AfterClass method that is invoked even for errors. deleteOnExit() does not work if the directory is non-empty (which can happen for errors)

**Pull request merged**